src/Tools/ProcessWatchdog/ProcessTracker.cs

```diff
@@ -0,0 +1,251 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Management;
+
+namespace ProcessWatchdog
+{
+    /// <summary>
+    /// Keeps track of a process and all its descendants.
+    /// </summary>
+    internal class ProcessTracker : IDisposable
+    {
+        private readonly Process _parentProcess;
+        private List<TrackedProcess> _trackedProcesses;
+        private readonly ProcDump _procDump;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProcessTracker"/> class from the
+        /// specified process id.
+        /// </summary>
+        /// <param name="parentProcess">
+        /// The process whose descendants are to be tracked.
+        /// </param>
+        /// <param name="procDump">
+        /// Object responsible for producing memory dumps of any tracked processes that
+        /// fail or are terminated.
+        /// </param>
+        internal ProcessTracker(Process parentProcess, ProcDump procDump)
+        {
+            _parentProcess = parentProcess;
+            _trackedProcesses = new List<TrackedProcess>();
+            _procDump = procDump;
+
+            TrackProcess(parentProcess);
+        }
+
+        internal bool AllFinished => !_trackedProcesses.Any();
+
+        internal void Update()
+        {
+            // Clear out any processes which have ended.
+            _trackedProcesses = _trackedProcesses.Where(tp => !tp.HasExited).ToList();
+
+            // CAUTION: This code is subject to a race condition where between one
+            // call to update and the next, all the processes in the list ended, but new
+            // processes (which we are not yet tracking) were created. In that case,
+            // _trackedProcesses would now be empty, and the ProcessWatchdog would exit,
+            // even though there are still processes we care about.
+            //
+            // This should not happen for the scenarios we care about, since the parent
+            // process should outlive all its descendants.
```

Comment:
This comment does not refer to the race condition I mentioned in the PR description. This is another, more benign race condition that (as I said in the comment here) I think doesn't actually occur in the scenario we care about.